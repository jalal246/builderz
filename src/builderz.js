import { rollup } from "rollup";
import packageSorter from "package-sorter";
import { getJsonByName, getJsonByPath } from "get-info";

import { getInput, getOutput } from "./config/index";

import { isValidArr, bindFunc } from "./utils";

import { SORT_PACKAGES, PKG_PATHS, PKG_NAMES } from "./constants";

import StateHandler from "./store";

/**
 * Write bundle.
 *
 * @param {Object} inputOptions
 * @param {Object} outputOptions
 */
async function build(inputFunc, outputFunc, { isProd, buildFormat, order }) {
  try {
    const input = await inputFunc({
      isProd,
      buildFormat,
      idx: order,
    });

    const output = await outputFunc({
      isProd,
      buildFormat,
    });

    /**
     * create a bundle
     */
    const bundle = await rollup(input);

    /**
     * write the bundle to disk
     */
    await bundle.write(output);
  } catch (err) {
    console.error(err);
  }
}

async function bundlePkg(state, pkgInfo, json, index) {
  let pkgPath;

  state.setNewPkg();

  if (typeof json === "string") {
    pkgPath = json;
  } else {
    /**
     * When name isn't valid, get-info assign index-order instead of name.
     */
    ({ path: pkgPath } = pkgInfo[json.name || index]);

    state.assignJson(json);
  }

  state.checkOpts();

  await state.setPkgPath(pkgPath);
  const bundleOpt = state.unpackBundleOpts();

  state.extractName();
  state.extractEntries();
  state.extractAlias();

  const inputFunc = bindFunc(getInput, state);
  const outputFunc = bindFunc(getOutput, state);
  const buildFunc = bindFunc(build, inputFunc, outputFunc);

  await Promise.all(bundleOpt.map(buildFunc));
}

async function sortAndBump(pkgJson, bundlePkgFunc) {
  const { sorted, unSorted } = packageSorter(pkgJson);

  if (isValidArr(unSorted)) {
    console.error(`Unable to sort packages: ${unSorted}`);
  }

  await sorted.reduce(async (sortedPromise, json) => {
    await sortedPromise;

    await bundlePkgFunc(json);
  }, Promise.resolve());
}

async function builderz(opts) {
  const state = new StateHandler(opts);

  const {
    [PKG_NAMES]: pkgNames = [],
    [PKG_PATHS]: pkgPaths = [],
    [SORT_PACKAGES]: sortPackages = true,
  } = state.generalOpts;

  const { json: pkgJson, pkgInfo = {}, unfoundJson } = isValidArr(pkgNames)
    ? getJsonByName(...pkgNames)
    : getJsonByPath(...pkgPaths);

  const bundlePkgFunc = bindFunc(bundlePkg, state, pkgInfo);

  try {
    if (isValidArr(unfoundJson)) {
      await Promise.all(unfoundJson.map((path, i) => bundlePkgFunc(path, i)));
    } else if (isValidArr(pkgJson)) {
      const isSequence =
        pkgJson.length > 1 && !isValidArr(unfoundJson) && sortPackages;

      if (isSequence) {
        await sortAndBump(pkgJson, bundlePkgFunc);
      } else {
        await Promise.all(pkgJson.map(bundlePkgFunc));
      }
    }
  } catch (err) {
    console.error(err);
  }
}

export default builderz;
