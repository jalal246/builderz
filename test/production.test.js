/* eslint-disable no-console */
import { resolve } from "path";
import { readdirSync, readFileSync } from "fs";
import del from "del";
import builderz from "../src";

describe("production", () => {
  it.each(["alias"])("tests %s)", async pkgName => {
    const pathPure = resolve(__dirname, "samples", pkgName);
    const distPath = resolve(pathPure, "dist");

    try {
      await builderz({
        isSilent: true,
        paths: [resolve(__dirname, pathPure)]
      });

      const files = readdirSync(distPath);
      expect(files.length).toMatchSnapshot();

      files
        .filter(file => !/\.map$/.test(file))
        .sort(file => (/modern/.test(file) ? 1 : 0))
        .forEach(file => {
          expect(
            readFileSync(resolve(distPath, file)).toString("utf8")
          ).toMatchSnapshot();
        });

      // await del(distPath);
    } catch (err) {
      console.error(err);
    }
  });
});
