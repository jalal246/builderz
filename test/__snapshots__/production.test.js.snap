// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`production tests pure js 1`] = `10`;

exports[`production tests pure js 2`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// this is currently broken because of a bug in terser
// https://github.com/terser-js/terser/issues/353

/** @__PURE__ */
function foo() {
  return \\"hello world\\";
}

exports.foo = foo;
"
`;

exports[`production tests pure js 3`] = `
"// this is currently broken because of a bug in terser
// https://github.com/terser-js/terser/issues/353

/** @__PURE__ */
function foo() {
  return \\"hello world\\";
}

export { foo };
"
`;

exports[`production tests pure js 4`] = `
"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0}),exports.foo=function(){return\\"hello world\\"};
//# sourceMappingURL=pure.min.cjs.js.map
"
`;

exports[`production tests pure js 5`] = `
"function o(){return\\"hello world\\"}export{o as foo};
//# sourceMappingURL=pure.min.esm.js.map
"
`;

exports[`production tests pure js 6`] = `
"!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?n(exports):\\"function\\"==typeof define&&define.amd?define([\\"exports\\"],n):n(((e=e||self).pure=e.pure||{},e.pure.min=e.pure.min||{},e.pure.min.umd=e.pure.min.umd||{},e.pure.min.umd.js={}))}(this,(function(e){\\"use strict\\";e.foo=function(){return\\"hello world\\"},Object.defineProperty(e,\\"__esModule\\",{value:!0})}));
//# sourceMappingURL=pure.min.umd.js.map
"
`;

exports[`production tests pure js 7`] = `
"(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = global || self, factory((global.pure = global.pure || {}, global.pure.umd = global.pure.umd || {}, global.pure.umd.js = {})));
}(this, (function (exports) { 'use strict';

  // this is currently broken because of a bug in terser
  // https://github.com/terser-js/terser/issues/353

  /** @__PURE__ */
  function foo() {
    return \\"hello world\\";
  }

  exports.foo = foo;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
//# sourceMappingURL=pure.umd.js.map
"
`;
